src/repr/type_binary.ml

```diff
@@ -18,6 +18,311 @@ open Type_core
+
+  type 'a shape_def = 'a -> t Unrolling.t
+
+  let get_size = function
+    | (`Int | `Int8 | `Int16 | `Int32 | `Int64) as n -> Boxed n
+    | `Fixed n -> Fixed n
+    | `Unboxed -> Any
+
+  let gen_placeholder_id : unit -> Uuid.t =
+    let counter = ref 0 in
+    fun () ->
+      incr counter;
+      string_of_int !counter
+
+  let rec of_type : type a. a ty shape_def =
+    let open Unrolling in
+    function
+    | Prim x -> return (prim x)
+    | Custom c -> return (custom c)
+    | Tuple x -> tuple x
+    | Option x ->
+        let+ x = of_type x in
+        Option x
+    | List x -> len_v x
+    | Array x -> len_v x
+    | Record r -> record r
+    | Variant v -> variant v
+    | Map m -> map m
+    | Boxed _ -> assert false
+    (* Recursive terms *)
+    | Self { self_unroll; _ } ->
+        let id = gen_placeholder_id () in
+        let placeholder : _ ty = Var id in
+        let+ inner =
+          Unrolling.with_rec_point id (of_type (self_unroll placeholder))
+        in
+        Recursive inner
+    | Var id -> (
+        let+ x = Unrolling.lookup id in
+        match x with
+        | None -> failwith "Malformed environment"
+        | Some (`Index i) -> Recur i)
+
+  and prim : type a. a prim -> t = function
+    | Unit -> Empty
+    | Bool -> Bool
+    | Char -> Char
+    | Int -> Int
+    | Int32 -> Int32
+    | Int64 -> Int64
+    | Float -> Float
+    | String len | Bytes len -> Contiguous (get_size len, Char)
+
+  and len_v : type a. a len_v shape_def =
+    let open Unrolling in
+    fun { len; v } ->
+      let+ v = of_type v in
+      Contiguous (get_size len, v)
+
+  and product : t list -> t =
+   fun components ->
+    if List.equal_elements shape_equal components then
+      Contiguous (Fixed (List.length components), List.hd components)
+    else Product components
+
+  and tuple : type a. a tuple shape_def =
+    let open Unrolling in
+    fun typ ->
+      let+ components =
+        match typ with
+        | Pair (a, b) ->
+            let* a = of_type a in
+            let+ b = of_type b in
+            [ a; b ]
+        | Triple (a, b, c) ->
+            let* a = of_type a in
+            let* b = of_type b in
+            let+ c = of_type c in
+            [ a; b; c ]
+      in
+      product components
+
+  and record : type a. a record shape_def =
+   fun { rfields = Fields (fs, _); _ } ->
+    let module Record_shape = Fields_folder (struct
+      type nonrec (_, _) t = t list Unrolling.t
+    end) in
+    let open Unrolling in
+    let nil = return [] in
+    let cons { ftype; _ } acc =
+      let* acc = acc in
+      let+ fshape = of_type ftype in
+      fshape :: acc
+    in
+    let+ components = Record_shape.fold { nil; cons } fs in
+    product components
+
+  and variant : type a. a variant shape_def =
+   fun v ->
+    let open Unrolling in
+    let+ cases =
+      Array.fold_right
+        (fun c acc ->
+          let* acc = acc in
+          match c with
+          | C0 _ -> return (Empty :: acc)
```

Comment:
If I am not mistaken, this means that the order of the elements in a `Repr.enum` is not reflected in a shape.

In general, a shape check cannot detect certain changes in the order of the cases and fields. This seems to be a problem.

Reply:
The problem with trying to distinguish those cases is that they're not actually distinguished in the underlying binary codec, so this becomes a bit of a slippery slope of trying to forbid certain "unwanted" but valid transformations on types. For instance, re-ordering a pair of fields of the same type is equivalent to renaming each one individually, so forbidding the former requires forbidding the latter (in order to keep transitivity on the shape equivalence relation). I can't think of a good way around that problem.

The current approach tries to stay away from such questions by only capturing information contained in the types themselves. If two fields of the same shape can't be permuted, the answer is probably to give them different shapes (by versioning their semantic interpretation with `Custom`).

Reply:
I was expecting the equivalence of shapes to imply the equivalence of the whole encode_bin/decode_bin process. I understand the subtle difference with the equivalent of codecs. I don't see a solution to this problem either.

Fortunately, such a change in the order of cases/fields should be fairly easy to detect during a code review.